### DISCUSSION TITLE
Is it possible to only match the longest literal string in a show rule?

### DISCUSSION THREAD

**Discussion by TS60** [typst/typst#4257] [Q&A] [answered]
Having the following show rules, I would prefer for some cases to only match the longest show rule and if there is a shorter one that also matches, only the longer one should match:

```typst
#show "going": "go-ing"
#show "ongoing": "on-going"

going

ongoing
```
This results in:
```
go-ing
on-go-ing
```

where I would prefer
```
go-ing
on-going
```

**Comment by laurmaedje** [comment, marked as answer]
You can turn both into one regex show rule matching on `regex("(on)?going")`. Since every rule matches only once, it won't rematch on its output.

You could also, in the second show rule, put some invisible zero width character somewhere into `going`, but that's far hackier.

**Reply by TS60** [reply]
Thanks, the first solution can solve it. The second one is exactly what I want to prevent because Typst would use the invisible zero width character as a hyphenation indicator, but I create the show rules to prevent hyphenation at non-word-boundaries, so this would not work for my use case.

The only problem with `regex` I see is that it probably requires N times p lookup time where p is the number of show rules. Where the non-regex solution probably uses aho corasick and thus is faster.

**Reply by laurmaedje** [reply]
I believe if you use ZWJ (zero width joiner) it shouldn't provide a hyphenation opportunity. ZWS would.

Regex can (but isn't yet) be optimized by compiling all active regexes into one finite automaton.

**Reply by TS60** [reply]
If this is true regarding "zero width joiner", this would be perfect, then your second solution would be the way to go. Maybe typst could even filter out these codepoints before including them in the pdf.

Ah, that makes sense regarding regex!

**Reply by TS60** [reply]
Yes, it works with ZWJ!

```typst
#show "going": "go-ing"
#show "ongoing": "on-goi‍ng" // with ZWJ
//#show "ongoing": "on-going" // without ZWJ

#set par(justify: true)
#set text(lang: "en", hyphenate: false)

going

#{
  for x in range(1000) {
    [ongoing ] + h(calc.rem(x, 10) * 1pt)
  }
}
```
